packages/teleport/src/Player/ProgressBar/ProgressBarDesktop.tsx

```diff
@@ -34,54 +35,94 @@ export const ProgressBarDesktop = (props: {
-              current: currentTimeMs,
-              time: toHuman(currentTimeMs),
-            };
+            const nextTimeMs = prevState.current + smoothingInterval;
+            if (nextTimeMs <= durationMs) {
+              return {
+                ...prevState,
+                current: nextTimeMs,
+                time: toHuman(nextTimeMs),
+              };
+            } else {
+              stopProgress();
+              return {
+                ...prevState,
+                current: durationMs,
+                time: toHuman(durationMs),
+              };
+            }
           });
+        }, smoothingInterval);
+      };
+
+      // The player always starts in play mode, so call this initially.
+      smoothOutProgress();
+
+      // Clears the smoothing interval and cancels any throttled updates,
+      // should be called when the playback is paused or ended.
+      const stopProgress = () => {
+        throttledUpdateCurrentTime.cancel();
+        clearInterval(intervalRef.current);
+      };
+
+      const throttledUpdateCurrentTime = throttle(
+        currentTimeMs => {
+          updateCurrentTime(currentTimeMs);
         },
-        // Magic number to throttle progress bar updates so that the playback is smoother.
+        // Magic number to throttle progress bar updates caused by TDP events
+        //  so that the playback is smoother.
         50
       );
 
+      // Listens for UPDATE_CURRENT_TIME events which coincide with
+      // TDP events sent to the playerClient by the server.
       playerClient.addListener(
         PlayerClientEvent.UPDATE_CURRENT_TIME,
         currentTimeMs => throttledUpdateCurrentTime(currentTimeMs)
       );
 
-      const progressToEnd = () => {
-        throttledUpdateCurrentTime.cancel();
-        // TODO(isaiah): Make this smoother
-        // https://github.com/gravitational/webapps/issues/579
+      playerClient.addListener(PlayerClientEvent.TOGGLE_PLAY_PAUSE, () => {
+        // setState({...state, isPlaying: !state.isPlaying}) doesn't work because
+        // the listener is added when state == initialState, and that initialState
+        // value is effectively hardcoded into its logic.
         setState(prevState => {
-          return { ...prevState, current: durationMs };
+          if (prevState.isPlaying) {
+            // pause
+            stopProgress();
+          } else {
+            // play
+            smoothOutProgress();
+          }
+          return { ...prevState, isPlaying: !prevState.isPlaying };
         });
-      };
-
-      playerClient.addListener(PlayerClientEvent.SESSION_END, () => {
-        progressToEnd();
       });
 
-      playerClient.addListener(TdpClientEvent.TDP_ERROR, () => {
-        progressToEnd();
+      playerClient.addListener(TdpClientEvent.WS_CLOSE, () => {
+        stopProgress();
       });
 
       return () => {
         throttledUpdateCurrentTime.cancel();
         playerClient.nuke();
+        stopProgress();
```

Comment:
line 123 no longer needed?

Reply:
It's still needed, that function is still called when we get events from the server (carrying TDP messages)

Reply:
even when stopProgress calls it already? :thinking:

Reply:
Oh no, I forgot that I'd added that. Thanks.